Data/ByteString/Char8.hs

```diff
@@ -1001,21 +1005,133 @@ unwords = intercalate (singleton ' ')
 readInt :: ByteString -> Maybe (Int, ByteString)
 {-# INLINABLE readInt #-}
 readInt bs = case B.uncons bs of
-    Just (w, rest) | w - 0x30 <= 9 -> readDec True bs    -- starts with digit
-                   | w == 0x2d     -> readDec False rest -- starts with minus
-                   | w == 0x2b     -> readDec True rest  -- starts with plus
-    _                              -> Nothing            -- not signed decimal
+    Just (w, rest) | w - 0x30 <= 9 -> cvtpos $ readWord64 bs
+                   | w == 0x2d     -> cvtneg $ readWord64 rest
+                   | w == 0x2b     -> cvtpos $ readWord64 rest
+    _                              -> Nothing
   where
-    -- | Read a decimal 'Int' without overflow.  The caller has already
+    -- | Conversion from Word64 to Int is value-preserving provided the input
+    -- does not exceed @'maxBound' :: 'Word'@, and the sign of the result
+    -- matches the sign of the input.  Overflow or underflow flips the sign.
+    cvtpos (Just (w, str))
+        | w <= wordMaxAsWord64
+        , let !i = fromIntegral w
+        , i >= 0 = Just (i, str)
```

Comment:
```suggestion
        | w <= intMaxAsWord64 = Just (fromIntegral w, str)
```
Seems a bit simpler, no?

I'd also suggest to introduce monomorphic helpers `word64ToInt = fromIntegral`, etc.

Reply:
I wanted to be sure to force `fromIntegral w`, so that ideally GHC would elide the thunk.  Hence `!i = ...`
Otherwise, you're right, your bounds check is equivalent, so I can drop one conditional. So it would be:

```
        | w <= intMaxAsWord64 = let !i = fromIntegral w in Just (i, str)
```